### ISSUE TITLE
Implicit gradient accumulation?

### ISSUE THREAD

**Issue by celsofranssa** [Lightning-AI/lightning#15045]
Hello, Lightning community,

Reading the documentation, I am still not sure about the behavior of using a dual (optimizer, scheduler) training as shown in the code snippet below:
```python
def configure_optimizers(self):
    # optimizers
    optimizer_1 = torch.optim.AdamW(self.encoder_1.parameters(), ...)
    optimizer_2 = torch.optim.AdamW(self.encoder_2.parameters(), ...)

    #schedulers 
    scheduler_1 = torch.optim.lr_scheduler.CyclicLR(optimizer_1, ...)
    scheduler_1 = torch.optim.lr_scheduler.CyclicLR(optimizer_1, ...)

    return (
        {"optimizer": optimizer_1,
         "lr_scheduler": {"scheduler": scheduler_1, "interval": "step", "name": "scheduler_1"},
         "frequency": 1},
        {"optimizer": optimizer_2,
         "lr_scheduler": {"scheduler": scheduler_2, "interval": "step", "name": "scheduler_2"},
         "frequency": 1},
    )
```
With `"frequency": 1` on both optimizers, the trainer calls  `optimizer_1.step()`  in step `i` while calling `optimizer_2.step()` in step `(i+1)`.

Therefore, does `optimizer_1.step()` only zero the gradient regarding `encoder_1` parameters? If so, was the gradient accumulated until the next step concerning `encoder_1` parameters?

**Comment by awaelchli** [comment]
Let's see if I get this right. 
I would say yes and no, it is up to what you are doing. 
Scenario 1 is: In step i you compute gradients for encoder 1, then optimizer 1 will step. Then in the step i+1 you compute gradients for encoder 2, and optimizer 2 will step.

Scenario 2: You compute both gradients for encoder 1 and 2, in this case on step i optimizer 1 will updated encoder 1 weights and zero grad them. In step i+1 you will again compute gradients for both but optimizer 2 will step and zero grad. This is probably undesired. 

If you need more control, I suggest to try [manual optimization](https://pytorch-lightning.readthedocs.io/en/stable/common/optimization.html#id2).

**Comment by celsofranssa** [comment]
@awaelchli,
Thank you for your answer. Using manual optimization, the `training_step` looks like this:

```python
    def training_step(self, batch, batch_idx):
        # samples and representations
        x_1, x_2 = batch["x"], batch["y"]
        x_rpr, y_rpr = self.encoder_1(x_1), self.encoder_2(x_2)

        # optimizers and schedulers
        optimizer_1, optimizer_2 = self.optimizers()
        scheduler_1, scheduler_2 = self.lr_schedulers()

        # loss and backward
        train_loss = self.loss(x_rpr, y_rpr)
        self.manual_backward(train_loss)

        
        if batch_idx % 2 == 0: # update encoder_1 in even steps
            optimizer_1.step()
            optimizer_1.zero_grad()
            scheduler_1.step()
        else: # update encoder_1 in odd steps
            optimizer_2.step()
            optimizer_2.zero_grad()
            scheduler_2.step()

        return train_loss
```
about which I interpret the following:

- in even iterations
-- only the `encoder_1`'s parameters are optimized;
-- the gradient of the `encoder_1`'s parameters is set to zero; and
-- the gradient of `encoder_2`'s parameters will be propagated to the next training iteration (gradient accumulation).

- in odd iterations:
-- only the `encoder_2`'s parameters are optimized;
-- the gradient of the `encoder_2`'s parameters is set to zero; and
-- the gradient of `encoder_1`'s parameters will be propagated to the next training iteration (gradient accumulation).

Is this interpretation correct?

**Comment by awaelchli** [comment]
Yes, I think that's correct. The question is just, is this really what you want there? Maybe consider adding `with torch.no_grad()` for optimizer 1 on the even steps and for optimizer 2 on odd steps.

**Comment by celsofranssa** [comment]
Sorry, I didn't get the effect of  `with torch.no_grad()` on the optimizers steps. Would you mind giving more details?

**Comment by awaelchli** [comment]
@celsofranssa You can choose if you want to "accumulate" or not. It is up to you, as you are in control of how to optimize your model. This means you can do
```py

if batch_idx % 2 == 0:
    # in this iteration, we will update the encoder 1
    x_rpr = self.encoder_1(x_1)

    with torch.no_grad():
         # and we won't need gradients for encoder 2
         y_rpr = self.encoder_2(x_2)

else:
    # in the odd iteration, it's in reverse

    y_rpr = self.encoder_2(x_2)
        with torch.no_grad():
             x_rpr = self.encoder_1(x_1)
        
```

or you can choose not to! It's just a matter of how you wan to use the two different optimizers.

**Comment by celsofranssa** [comment]
I got it, thank you!
